### PR TITLE
SDK2: Ensure service endpoints track2

### PR DESCRIPTION
**Pull request by bitoku**
### Which issue this PR addresses:

Fixes https://issues.redhat.com/browse/ARO-4665 and https://issues.redhat.com/browse/ARO-7316

### What this PR does / why we need it:

This PR replaces old SDK to the new one.

### Test plan for issue:

ci, e2e

### Is there any documentation that needs to be updated for this PR?

no

### How do you know this will function as expected in production? 

Creating a cluster can verify the change.
It can be checked in e2e.

### CODE REVIEW

File: pkg/cluster/ensureendpoints_test.go [hunk 1]
```diff
@@ -19,88 +23,438 @@ var (
...
+func TestAddEndpointsToSubnets(t *testing.T) {
+	for _, tt := range []struct {
+		name           string
+		subnet         *armnetwork.Subnet
+		newEndpoints   []string
+		expectedSubnet *armnetwork.Subnet
+		expectedResult bool
+	}{
+		{
+			name: "addEndpointsToSubnet should  when the subnet contains all new endpoints in succeeded state",
```
Comment (kimorris27) [inline pkg/cluster/ensureendpoints_test.go:282]: Nit:

```suggestion
			name: "addEndpointsToSubnet should do nothing when the subnet contains all new endpoints in succeeded state",
```

**Review by kimorris27 — APPROVED** [review]
LGTM except for a tiny nit.

**Comment by kimorris27** [comment]
One question though: how does this address the cluster install failure?

**Comment by bitoku** [comment]
@kimorris27 

We have seen some install failures about authorization in this step.
It should be caused by the AAD propagation delay.

We can wrap the step with `AuthorizationRetryingAction`, but it will retry the whole step again.
Because we don't make steps idempotent, retrying the whole step might cause different issues.

By updating the old SDK to the new one, we can use built-in retry strategy in the new SDK.
This will retry just the failed API, so it's a safer way than `AuthorizationRetryingAction`.

Aside from the CIF, we should update old SDKs anyway.

**Review by kimorris27 — APPROVED** [review]
> @kimorris27
> 
> We have seen some install failures about authorization in this step. It should be caused by the AAD propagation delay.
> 
> We can wrap the step with `AuthorizationRetryingAction`, but it will retry the whole step again. Because we don't make steps idempotent, retrying the whole step might cause different issues.
> 
> By updating the old SDK to the new one, we can use built-in retry strategy in the new SDK. This will retry just the failed API, so it's a safer way than `AuthorizationRetryingAction`.
> 
> Aside from the CIF, we should update old SDKs anyway.

Oh, cool. I didn't realize the new SDK clients had that built-in retry strategy. That makes sense!

File: pkg/cluster/ensureendpoints.go [hunk 2]
```diff
@@ -36,3 +64,49 @@ func (m *manager) getSubnetIds() ([]string, error) {
...
 	return subnets, nil
 }
+
+// addEndpointsToSubnet adds the endpoints (that either are missing in subnet
+// or aren't in succeeded state in the subnet) to the subnet and returns the updated subnet
+func addEndpointsToSubnet(endpoints []string, subnet *armnetwork.Subnet) (subnetChanged bool) {
+	for _, endpoint := range endpoints {
+		endpointFound, serviceEndpointPtr := subnetContainsEndpoint(subnet, endpoint)
+
+		if !endpointFound || *serviceEndpointPtr.ProvisioningState != armnetwork.ProvisioningStateSucceeded {
```
Comment (dkeohane) [inline pkg/cluster/ensureendpoints.go:74]: Can ProvisioningState ever be in an error state? We wouldn't want to add it then I'm guessing

**Review by dkeohane — COMMENTED** [review]
Left a quick review, didn't look through unit tests added.

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -98,11 +98,12 @@ type manager struct {
 	fpPrivateEndpoints       network.PrivateEndpointsClient // TODO: use armFPPrivateEndpoints instead.
 	armFPPrivateEndpoints    armnetwork.PrivateEndpointsClient
 	armRPPrivateLinkServices armnetwork.PrivateLinkServicesClient
+	armSubnets               armnetwork.SubnetsClient
 	userAssignedIdentities   armmsi.UserAssignedIdentitiesClient
 
 	dns     dns.Manager
 	storage storage.Manager
-	subnet  subnet.Manager
+	subnet  subnet.Manager // TODO: use armSubnet instead.
```
[the remark this answers is not in the document]
Comment (bitoku) [reply pkg/cluster/cluster.go:106]: It should be removed when we get rid of all usages of old SDKs.
It's a part of https://issues.redhat.com/browse/ARO-4665 so I don't think we card it up.

File: pkg/cluster/ensureendpoints.go (hunk 2 above)
Comment (bitoku) [reply pkg/cluster/ensureendpoints.go:74]: This is a copy/paste from the old code.
Also this function adds endpoints that are not available in the subnet, so it's supposed to be ok.

**Review by jaitaiwan — APPROVED** [review]
LGTM. 1 small nit

**Review by kimorris27 — APPROVED** [review]

**Review by UlrichSchlueter — APPROVED** [review]
Code LGTM

Has this been tested against a private cluster ?

**Comment by bitoku** [comment]
not yet. I'll test it in local dev rp.

**Comment by UlrichSchlueter** [comment]
@bitoku never mind, I misread one of the comments

**Comment by bitoku** [comment]
FYI, it worked for a private cluster as well.

**Pull request merged**